sw/host/opentitanlib/src/io/jtag.rs

```diff
@@ -0,0 +1,92 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::Result;
+use serde::{Deserialize, Serialize};
+use structopt::StructOpt;
+use thiserror::Error;
+
+use std::path::PathBuf;
+use std::rc::Rc;
+
+use crate::app::TransportWrapper;
+use crate::dif::lc_ctrl::LcCtrlReg;
+use crate::impl_serializable_error;
+
+#[derive(Debug, StructOpt, Clone)]
+pub struct JtagParams {
+    /// OpenOCD binary path.
+    #[structopt(long, default_value = "openocd")]
+    pub openocd: PathBuf,
+
+    /// Path to OpenOCD JTAG adapter config file.
+    #[structopt(long)]
+    pub openocd_adapter_config: Option<String>,
+
+    /// Path to OpenOCD JTAG target config file for the RISC-V TAP.
+    #[structopt(long)]
+    pub openocd_riscv_target_config: Option<String>,
+
+    /// Path to OpenOCD JTAG target config file for the LC TAP.
+    #[structopt(long)]
+    pub openocd_lc_target_config: Option<String>,
+}
+
+impl JtagParams {
+    pub fn create(&self, transport: &TransportWrapper) -> Result<Rc<dyn Jtag>> {
+        let jtag = transport.jtag(&self)?;
+        Ok(jtag)
+    }
+}
+
+/// Errors related to the JTAG interface.
+#[derive(Error, Debug, Deserialize, Serialize)]
+pub enum JtagError {
+    #[error("JTAG timeout")]
+    Timeout,
+    #[error("JTAG busy")]
+    Busy,
+    #[error("Generic error {0}")]
+    Generic(String),
+}
+impl_serializable_error!(JtagError);
+
+/// A trait which represents a JTAG interface.
+pub trait Jtag {
+    /// Connect to the given JTAG TAP.
+    fn connect(&self, tap: JtagTap) -> Result<()>;
+    /// Disconnect from the TAP.
+    fn disconnect(&self) -> Result<()>;
+
+    // Commands
+
+    /// Read a lifecycle controller register.
+    fn read_lc_ctrl_reg(&self, reg: &LcCtrlReg) -> Result<u32>;
+
+    /// Write a value to a lifecycle controller register.
+    fn write_lc_ctrl_reg(&self, reg: &LcCtrlReg, value: u32) -> Result<()>;
```

Comment:
I think we just want:
```suggestion
    fn read_reg(&self, reg: ...) -> Result<u32>;
    fn write_reg(&self, reg: ..., value: u32) -> Result<()>;
```
because we want to be able to read/write registers in both the CPU and lc_ctrl depending on what TAP is selected. Or was your thought to add `{read,write}_cpu_reg()` functions later too? that only work when the CPU TAP is selected?

Regardless, let's just leave these for now and work out the API in a follow up PR. I want to get this merged so we can iterate on this.

Reply:
Oh yes, these should be in the API too, I just missed them out since I wasn't testing the CPU TAP. Thanks for catching.

Reply:
Leaving this for a future PR since we'll need some kind of `IbexReg`